fms_extras/utils/generation.py

```diff
@@ -3,9 +3,623 @@
+        n_candidates: only consider the top n most confident candidates from the speculator
+        threshes: build candidate suffix trees by taking top-(threshes[n]) most confident values
+            for head n. len(threshes) must equal speculator.n_predict. prod(threshes) must be greater
+            than or equal to n_candidates (we cannot have more candidates than tree leaves).
+        flattening: enable batch flattening / tree attention with redundant prefix removal when compression
+            ratio is favorable. Adds extra overhead to shrink the token count in each batch.
+        decode_model: nn.Module, optional
+            an optional model that performs the decode forward step. If None, will use the function or nn.Module
+            provided to model param to perform decode forward. This parameter is intended to be used when the compile
+            flags for the model doing prefill and the model generating subsequent tokens are different (default is None)
+        cudagraphs: bool
+            if True, cudagraphs is used and all metadata will be padded, otherwise metadata will not be padded unless
+            required. Note: This is a WIP and only works for batch_size=1
+    Returns:
+        result: List of id tensors, possibly different lengths if batching.
+        n_steps: Number of foward passes used to generate provided tokens.
+    """
+    start_time = time.time()
+    # Construct batch(es) and initial inputs
+    if decode_model is None:
+        decode_model = model
+    bsize = len(input_ids)
+
+    if cudagraphs and (flattening or bsize != 1):
+        raise NotImplementedError(
+            "cudagraphs is not yet supported for batch sizes greater than 1 or flatting"
+        )
+
+    result = input_ids  # [b] n
+    inp_len = speculator.n_predict + 1
+    prompt_lengths = [min(seq.size(0), max_seq_len) for seq in input_ids]
+
+    # reserve blocks in the cache for the prompts for prefill
+    cache_data: PagedAttentionCacheData = kv_cache_manager.allocate_tokens(
+        prompt_lengths
+    )
+
+    # execute the prefill step
+    # logits -  b 1 d
+    # embeds - b 1 v
+    logits, embeds = __execute_prefill(
+        model, input_ids, cache_data, max_seq_len - inp_len
+    )
+
+    # get the time to first token
+    ttft = time.time() - start_time
+
+    # setting local variables for re-use prior to decode loop
+    start_time = time.time()
+    parent_sequence_ids = cache_data.sequence_ids
+    n_gen = [0] * bsize
+    n_steps = 0
+    input_ids = torch.argmax(logits, dim=2)  # b 1
+    result = [
+        torch.cat((line, input_id), dim=0)
+        for line, input_id in zip(result, list(input_ids))
+    ]
+    block_mapping_max = ((max(prompt_lengths) + new_tokens - 1) // 16) + 1
+    model_input_lengths = [inp_len] * (input_ids.size(0) * n_candidates)
+
+    # perform decode step
+    while min(n_gen) < new_tokens:
+        n_steps += 1
+
+        ############### Prepare Inputs for decode ###############
```

Comment:
these block comments dividing (what was a much longer piece of) code into logical sections seems less necessary as we condense each sections each into 1-2 function calls (especially if we can make the function names basically match whatever would be the best comment comment). It might be better to remove them so more of the logic can be seen on one page

Reply:
fixed in latest commit